# Exercise 4.7 Deriving New Variables for IC Data

This script contains

1. Import libraries
2. Create price_label variable
3. Create busiest_day variable
4. Create busiest_days variable
5. Create busiest_period_of_day variable
6. Export as pkl file

# 01. Importing Libraries

In [7]:
# Importing libraries

import pandas as pd
import numpy as np
import os

In [8]:
# Path to IC folders
path = r"C:\Users\alabaro\OneDrive - DAI\Documents\Aug_2023 Instacart Basket Analysis"

In [9]:
# importing orders_products_merged
df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

# 02. Create price_label Variable

In [10]:
# create a subset

df = df_ords_prods_merged[:1000000]

In [11]:
# check output

df.shape

(1000000, 15)

In [12]:
# define a function for new price labels

def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [13]:
# apply new function

df['price_range'] = df.apply(price_label, axis=1)

C:\Users\alabaro\AppData\Local\Temp\ipykernel_20308\1271307185.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [14]:
# check counts for each new price range

df['price_range'].value_counts(dropna=False)

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [15]:
# check max price

df['prices'].max()

14.8

Maximum price in this subset is 14.8, which confirms why there are 0 High-range products. 

# 03. Create 'busiest_day' Variable

In [18]:
# check output

df_ords_prods_merged.head()

order_id  user_id eval_set  order_number  order_dow  order_hour_of_day  \
0   2539329        1    prior             1          2                  8   
1   2398795        1    prior             2          3                  7   
2    473747        1    prior             3          3                 12   
3   2254736        1    prior             4          4                  7   
4    431534        1    prior             5          4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

In [21]:
# Rename order_dow to order_days_of_week
df_ords_prods_merged.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [23]:
# check counts for order_day_of_week

df_ords_prods_merged['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

Saturday is the busiest day and Wednesday is the slowest day

In [25]:
# results for new column 'busiest_day'

result = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [26]:
df_ords_prods_merged['busiest_day'] = result

In [27]:
df_ords_prods_merged['busiest_day'].value_counts(dropna=False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

The frequency counts above for orders_day_of_week match with the Busiest day (0 for Saturday - 6204182) and Least busy (4 for Wednesday - 3783802)

# 04. Create 'busiest_days' variable

In [28]:
# check counts for order_day_of_week

df_ords_prods_merged['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

2 busiest days are Saturday and Sunday, 2 least busy days are Tuesday annd Wednesday

In [39]:
# Create new if-statement for-loop for new requirements

result_v2 = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_v2.append("Busiest days")
  elif value == 4 or value == 3:
    result_v2.append("Slowest days")
  else:
    result_v2.append("Regularly busy")

In [40]:
# Add result to dataframe in new column

df_ords_prods_merged['busiest_days'] = result_v2

In [44]:
# check output

df_ords_prods_merged.head(5)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

      busiest_day    busiest_days  
0  Regularly busy  Regularly busy  
1  Regularly busy    Slowest days  
2  Regularly busy    Slowest days  
3      Least busy    Slowest days  
4      Least busy    Slowest days

In [45]:
#check frequency

df_ords_prods_merged['busiest_days'].value_counts(dropna=False)

busiest_days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

Checking output by adding together the busiest and slowest days. 
Busiest days = 11864412. Saturday (6204182) + Sunday (5660230) = 11864412 | Values match
Slowest days = 7624336. Wednesday (3783802) + Tuesday (3840534) = 7624336 | values match

# 05. Create 'busiest_period_of_day' variable

In [46]:
# check counts for order_hour_of_day

df_ords_prods_merged['order_hour_of_day'].value_counts(dropna=False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

We can divide the hours into three different groups
Most orders - 10, 11, 14, 15, 13, 12, 16, 9
Average orders - 17, 8, 18, 19, 20, 7, 21, 22
Fewest orders - 23, 6, 0, 1, 5, 2, 4, 3 

In [47]:
# Create new if-statement for-loop for new requirements

result_v3 = []

for value in df_ords_prods_merged["order_hour_of_day"]:
  if value == 10 or value == 11 or value == 14 or value == 15 or value == 13 or value == 12 or value == 16 or value == 9:
    result_v3.append("Most orders")
  elif value == 17 or value == 8 or value == 18 or value == 19 or value == 20 or value == 7 or value == 21 or value == 22:
    result_v3.append("Average orders")
  else:
    result_v3.append("Fewest orders")

In [48]:
# Add result to dataframe in new column

df_ords_prods_merged['busiest_period_of_day'] = result_v3

In [49]:
# check output

df_ords_prods_merged.head(5)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

      busiest_day    busiest_days busiest_period_of_day  
0  Regularly busy  Regularly busy        Average orders  
1  Regularly busy    Slowest days        Average orders  
2  Regularly busy    Slowest days           Most orders  
3      Least busy    Slowest days        Average orders  
4      Least busy    Slowest days           Most orders

In [50]:
# check frequency

df_ords_prods_merged['busiest_period_of_day'].value_counts(dropna=False)

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

# 06. Export as pkl

In [51]:
df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged_v2.pkl'))